lua/auto-session-autocmds.lua

```diff
@@ -12,46 +32,70 @@ M.setup_autocmds = function(config, AutoSession)
   end
 
   local conf = config.cwd_change_handling
+  local scopes = { "global", "tabpage"}
 
-  vim.api.nvim_create_autocmd("DirChangedPre", {
-    callback = function()
-      Lib.logger.debug "DirChangedPre"
-      Lib.logger.debug("cwd: " .. vim.fn.getcwd())
+  for _, pattern in ipairs(scopes) do
+    vim.api.nvim_create_autocmd("DirChangedPre", {
+      callback = function()
+        Lib.logger.debug "DirChangedPre"
+        Lib.logger.debug("  cwd: " .. vim.fn.getcwd())
+        Lib.logger.debug("  target: " .. vim.v.event.directory)
+        Lib.logger.debug("  changed window: " .. tostring(vim.v.event.changed_window))
+        Lib.logger.debug("  scope: " .. vim.v.event.scope)
 
-      AutoSession.AutoSaveSession()
+        -- Don't want to save session if dir change was triggered
+        -- by a window change. This will corrupt the session data,
+        -- mixing the two different directory sessions
+        if vim.v.event.changed_window then
+          return
+        end
 
-      -- Clear all buffers and jumps after session save so session doesn't blead over to next session.
-      vim.cmd "%bd!"
-      vim.cmd "clearjumps"
+        AutoSession.AutoSaveSession()
 
-      if type(conf.pre_cwd_changed_hook) == "function" then
-        conf.pre_cwd_changed_hook()
-      end
-    end,
-    pattern = "global",
-  })
+        -- Clear all buffers and jumps after session save so session doesn't blead over to next session.
+        vim.cmd "%bd!"
+        vim.cmd "clearjumps"
 
-  if conf.restore_upcoming_session then
-    vim.api.nvim_create_autocmd("DirChanged", {
-      callback = function()
-        Lib.logger.debug "DirChanged"
-        Lib.logger.debug("cwd: " .. vim.fn.getcwd())
+        if type(conf.pre_cwd_changed_hook) == "function" then
+          conf.pre_cwd_changed_hook()
+        end
+      end,
+      pattern = pattern,
+    })
 
-        -- Deferring to avoid otherwise there are tresitter highlighting issues
-        vim.defer_fn(function()
```

Comment:
This approach was causing an infinite loop. From the docs:

>DirChanged			After the [current-directory](https://neovim.io/doc/user/editing.html#current-directory) was changed.
> ...
>				*Non-recursive (event cannot trigger itself).*

By triggering `Autosession.AutoRestoreSession` asynchronously, it caused an infinite loop where the opened buffer could change the global directory (somehow) after opening. So, I moved the asynchronous part to after the session was restored. It then runs `filetype detect` on all the open buffers which fixes the syntax highlighting error.